Mathlib/Probability/Kernel/Disintegration/Basic.lean

```diff
@@ -0,0 +1,518 @@
+  rw [condKernelReal, compProd_toKernel]
+
+/-- Auxiliary definition for `ProbabilityTheory.kernel.condKernel`.
+A conditional kernel for `κ : kernel α (γ × Ω)` where `γ` is countably generated and `Ω` is
+standard Borel. -/
+noncomputable
+def condKernelBorel (κ : kernel α (γ × Ω)) [IsFiniteKernel κ] : kernel (α × γ) Ω :=
+  let e := embeddingReal Ω
+  let he := measurableEmbedding_embeddingReal Ω
+  let κ' := map κ (Prod.map (id : γ → γ) e) (measurable_id.prod_map he.measurable)
+  borelMarkovFromReal Ω (condKernelReal κ')
+
+instance instIsMarkovKernelCondKernelBorel (κ : kernel α (γ × Ω)) [IsFiniteKernel κ] :
+    IsMarkovKernel (condKernelBorel κ) := by
+  rw [condKernelBorel]
+  infer_instance
+
+lemma compProd_fst_condKernelBorel (κ : kernel α (γ × Ω)) [IsFiniteKernel κ] :
+    fst κ ⊗ₖ condKernelBorel κ = κ := by
+  rw [condKernelBorel, compProd_fst_borelMarkovFromReal _ _ (compProd_fst_condKernelReal _)]
+
+end CountablyGenerated
+
+section Unit
+
+/-- Auxiliary definition for `MeasureTheory.Measure.condKernel` and
+`ProbabilityTheory.kernel.condKernel`.
+A conditional kernel for `κ : kernel Unit (α × ℝ)`. -/
+noncomputable
+def condKernelUnitReal (κ : kernel Unit (α × ℝ)) [IsFiniteKernel κ] : kernel (Unit × α) ℝ :=
+  (isCondKernelCDF_condCDF (κ ())).toKernel
+
+instance instIsMarkovKernelCondKernelUnitReal (κ : kernel Unit (α × ℝ)) [IsFiniteKernel κ] :
+    IsMarkovKernel (condKernelUnitReal κ) := by
+  rw [condKernelUnitReal]
+  infer_instance
+
+lemma compProd_fst_condKernelUnitReal (κ : kernel Unit (α × ℝ)) [IsFiniteKernel κ] :
+    fst κ ⊗ₖ condKernelUnitReal κ = κ := by
+  rw [condKernelUnitReal, compProd_toKernel]
+  ext a
+  simp
+
+/-- Auxiliary definition for `MeasureTheory.Measure.condKernel` and
+`ProbabilityTheory.kernel.condKernel`.
+A conditional kernel for `κ : kernel Unit (α × Ω)` where `Ω` is standard Borel. -/
+noncomputable
+def condKernelUnitBorel (κ : kernel Unit (α × Ω)) [IsFiniteKernel κ] : kernel (Unit × α) Ω :=
+  let e := embeddingReal Ω
+  let he := measurableEmbedding_embeddingReal Ω
+  let κ' := map κ (Prod.map (id : α → α) e) (measurable_id.prod_map he.measurable)
+  borelMarkovFromReal Ω (condKernelUnitReal κ')
+
+instance instIsMarkovKernelCondKernelUnitBorel (κ : kernel Unit (α × Ω)) [IsFiniteKernel κ] :
+    IsMarkovKernel (condKernelUnitBorel κ) := by
+  rw [condKernelUnitBorel]
+  infer_instance
+
+lemma compProd_fst_condKernelUnitBorel (κ : kernel Unit (α × Ω)) [IsFiniteKernel κ] :
+    fst κ ⊗ₖ condKernelUnitBorel κ = κ := by
+  rw [condKernelUnitBorel,
+    compProd_fst_borelMarkovFromReal _ _ (compProd_fst_condKernelUnitReal _)]
+
+end Unit
+
+section Measure
+
+variable {ρ : Measure (α × Ω)} [IsFiniteMeasure ρ]
+
+/-- Conditional kernel of a measure on a product space: a Markov kernel such that
+`ρ = ρ.fst ⊗ₘ ρ.condKernel` (see `MeasureTheory.Measure.compProd_fst_condKernel`). -/
+noncomputable
+def _root_.MeasureTheory.Measure.condKernel (ρ : Measure (α × Ω)) [IsFiniteMeasure ρ] :
```

Comment:
`irreducible_def`, maybe, to hide the messy details of the construction (which should never be relevant once the API is set up)?